File: core/inpainter/policy.py
from __future__ import annotations

from typing import Any

from inpainter.paths import policy_dir
from inpainter.schema import apply_defaults
from inpainter.state import load_state, persist_state
from inpainter.ts_policy import run_transition


def transition(state: Any, event: dict[str, Any]) -> dict[str, Any]:
    return run_transition(policy_dir(), "global", state, event)


def dispatch(event: dict[str, Any]) -> dict[str, Any]:
    current = load_state()
    result = transition(current, event)
    next_state = result.get("state")
    if not isinstance(next_state, str) or not next_state:
        next_state = current or "unknown"
    persist_state(next_state)
    payload = result.get("payload")
    if not isinstance(payload, dict):
        payload = {}
    reported = apply_defaults("status", payload)
    reported["state"] = next_state
    return reported

File: core/src/policy.ts
import { join } from "node:path";

import { createCorePolicy } from "@inpainter/core-policy";
import type { PolicyEvent } from "@inpainter/policy-runtime";

import { coreRoot } from "./paths.ts";
import { applyDefaults } from "./schema.ts";
import { loadState, persistState } from "./state.ts";

type CorePolicy = Awaited<ReturnType<typeof createCorePolicy>>;

let policy: CorePolicy | undefined;

async function loadedPolicy(): Promise<CorePolicy> {
  if (!policy) {
    policy = await createCorePolicy(join(coreRoot(), "scripts"));
  }
  return policy;
}

export async function transition(state: unknown, event: PolicyEvent) {
  return (await loadedPolicy()).dispatch(state, event);
}

export async function dispatch(event: PolicyEvent): Promise<Record<string, unknown>> {
  const current = loadState();
  const result = await transition(current, event);
  const nextState =
    typeof result.state === "string" && result.state ? result.state : (current ?? "unknown");
  persistState(nextState);
  const payload =
    result.payload && typeof result.payload === "object" && !Array.isArray(result.payload)
      ? result.payload
      : {};
  const reported = applyDefaults("status", payload);
  reported.state = nextState;
  return reported;
}
